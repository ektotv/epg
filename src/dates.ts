export function xmltvTimestampToUtcDate(timestamp: string) {
  const cleanTimestamp = timestamp.replace(/[- :Zz]/g, "");

  const year = parseInt(cleanTimestamp.slice(0, 4), 10);
  const month = parseInt(cleanTimestamp.slice(4, 6), 10) - 1 || 0;
  const day = parseInt(cleanTimestamp.slice(6, 8), 10) || 1;
  const hours = parseInt(cleanTimestamp.slice(8, 10), 10) || 0;
  const minutes = parseInt(cleanTimestamp.slice(10, 12), 10) || 0;
  const seconds = parseInt(cleanTimestamp.slice(12, 14), 10) || 0;
  const timeZone = cleanTimestamp.length > 14 ? cleanTimestamp.slice(-5) : null;

  const date = new Date(Date.UTC(year, month, day, hours, minutes, seconds));

  if (timeZone) {
    const offsetSign = timeZone.slice(0, 1);
    const offsetHours = parseInt(timeZone.slice(1, 3), 10);
    const offsetMinutes = parseInt(timeZone.slice(3, 5), 10);
    const totalOffsetMinutes = offsetHours * 60 + offsetMinutes;

    if (offsetSign === "+") {
      date.setUTCMinutes(date.getUTCMinutes() - totalOffsetMinutes);
    } else if (offsetSign === "-") {
      date.setUTCMinutes(date.getUTCMinutes() + totalOffsetMinutes);
    }
  }

  return date;
}

export function dateToXmltvUtcTimestamp(date: Date) {
  const padZeroes = (num: number) => (num < 10 ? `0${num}` : num.toString());

  const year = padZeroes(date.getUTCFullYear());
  const month = padZeroes(date.getUTCMonth() + 1);
  const day = padZeroes(date.getUTCDate());
  const hours = padZeroes(date.getUTCHours());
  const minutes = padZeroes(date.getUTCMinutes());
  const seconds = padZeroes(date.getUTCSeconds());

  return `${year}${month}${day}${hours}${minutes}${seconds} +0000`;
}
