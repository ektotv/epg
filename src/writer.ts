import type { XmltvDOM, XmltvDOMNode } from "./types";

export function writer(xmltvDom: XmltvDOM): string {
  let out = "";
  function writeChildren(node: XmltvDOM) {
    if (node)
      for (var i = 0; i < node.length; i++) {
        if (typeof node[i] === "string") {
          if ((node[i] as string).includes("!DOCTYPE")) {
            out += "<" + (node[i] as string).trim() + ">";
            continue;
          }
          out += (node[i] as string).trim();
        } else {
          writeNode(node[i]);
        }
      }
  }
  function writeNode(node: XmltvDOMNode) {
    if (typeof node === "string") return;
    out += "<" + node.tagName;
    for (const attr in node.attributes) {
      if (
        typeof node.attributes[attr] === "string" &&
        node.attributes[attr].indexOf('"') === -1
      ) {
        out += " " + attr + '="' + node.attributes[attr].trim() + '"';
      } else if (typeof node.attributes[attr] === "boolean") {
        out += " " + attr + '="' + (node.attributes[attr] ? "yes" : "no") + '"';
      } else {
        out += " " + attr + "='" + node.attributes[attr] + "'";
      }
    }

    if (["new", "icon", "previously-shown"].indexOf(node.tagName) >= 0) {
      out += "/>";
      return;
    }

    if (node.tagName === "?xml") {
      out += "?>";
      return;
    }
    out += ">";
    if (typeof node.children === "boolean") {
      out += node.children ? "yes" : "no";
      return;
    }
    writeChildren(node.children);
    out += "</" + node.tagName + ">";
  }

  writeChildren(xmltvDom);

  let header = "";

  if (out.indexOf("?xml") === -1) {
    header += '<?xml version="1.0" encoding="UTF-8"?>';
  }
  if (out.indexOf("!DOCTYPE") === -1) {
    header += '<!DOCTYPE tv SYSTEM "xmltv.dtd">';
  }

  return header + out;
}
