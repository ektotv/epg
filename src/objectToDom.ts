import type { XmltvDOMNode } from "./types";
import { dateToXmltvUtcTimestamp } from "./dates.js";
import {
  ReversedXmlTranslations,
  xmltvTranslationsReversed,
  xmltvTagAttributes,
} from "./xmltvTagsAttributes.js";

/**
 * Converts an XMLTV object to a DOM tree
 *
 * @param obj The XMLTV object to convert to a DOM tree
 * @param key The current key to loop over
 * @param isArrayChild Controls if the return is an array or not
 * @returns The DOM tree
 */
export function objectToDom(obj: any, key = "tv", isArrayChild = false): any {
  if (Array.isArray(obj)) {
    return obj.map((item) => objectToDom(item, key, true));
  }

  if (typeof obj === "number") {
    return obj.toString();
  }

  if (typeof obj === "string") {
    return obj;
  }

  if (obj instanceof Date && key !== "date") {
    return dateToXmltvUtcTimestamp(obj);
  }

  if (typeof obj === "boolean" && key !== "new") {
    return obj ? "yes" : "no";
  }

  const translatedTagName =
    xmltvTranslationsReversed.tags[
      key as keyof ReversedXmlTranslations["tags"]
    ] || key;

  const xmltvDOMNode: XmltvDOMNode = {
    tagName: translatedTagName,
    attributes: {},
    children: [],
  };

  for (let childKey in obj) {
    const translatedAttributeName =
      xmltvTranslationsReversed.attributes[
        childKey as keyof ReversedXmlTranslations["attributes"]
      ] || childKey;

    if (obj[childKey].tagName === "new") {
      obj[translatedTagName].children = [];
      continue;
    }

    if (
      (xmltvTagAttributes.indexOf(translatedAttributeName) >= 0 &&
        typeof obj[childKey] !== "object") ||
      obj[childKey] instanceof Date
    ) {
      if (xmltvDOMNode.tagName === "credits" && childKey === "guest") {
        continue;
      }
      if (xmltvDOMNode.tagName === "programme" && childKey === "channel") {
        xmltvDOMNode.attributes[translatedAttributeName] = obj[childKey];
        continue;
      }

      if (xmltvDOMNode.tagName === "tv" && childKey === "date") {
        xmltvDOMNode.attributes[translatedAttributeName] =
          dateToXmltvUtcTimestamp(obj[childKey]);
        continue;
      }

      if (xmltvDOMNode.tagName === "programme" && childKey === "date") {
        xmltvDOMNode.children.push({
          tagName: translatedAttributeName,
          attributes: {},
          children: [dateToXmltvUtcTimestamp(obj[childKey])],
        });
        continue;
      }

      const childJsType = typeof obj[childKey];

      if (childJsType === "number") {
        xmltvDOMNode.attributes[translatedAttributeName] =
          obj[childKey].toString();
        continue;
      }

      if (childJsType === "boolean") {
        xmltvDOMNode.attributes[translatedAttributeName] = obj[childKey]
          ? "yes"
          : "no";
        continue;
      }

      if (childJsType === "object" && !Array.isArray(obj[childKey])) {
        if (obj[childKey] instanceof Date) {
          obj[childKey] = dateToXmltvUtcTimestamp(obj[childKey]);
          xmltvDOMNode.attributes[translatedAttributeName] = obj[childKey];
          continue;
        }

        const normalizedAttrs = Object.keys(obj[childKey]).map((key) => {
          obj[childKey][key] = obj[childKey][key].toString();
        });

        xmltvDOMNode.attributes[translatedAttributeName] = normalizedAttrs;
        continue;
      }
      xmltvDOMNode.attributes[translatedAttributeName] = obj[childKey];
    } else {
      const childNode = objectToDom(obj[childKey], childKey);
      if (Array.isArray(childNode)) {
        xmltvDOMNode.children = [...xmltvDOMNode.children, ...childNode];
      } else {
        if (childKey !== "_value") {
          xmltvDOMNode.children.push({
            tagName: translatedAttributeName,
            attributes: {},
            children: [childNode],
          });
          continue;
        }
        xmltvDOMNode.children.push(childNode);
      }
    }
  }

  return isArrayChild ? xmltvDOMNode : [xmltvDOMNode];
}
