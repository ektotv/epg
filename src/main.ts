import { parser } from "./parser.js";
import { writer } from "./writer.js";
import { objectToDom } from "./objectToDom.js";
import { toObject } from "./toObject.js";
import type {
  Xmltv,
  XmltvAudio,
  XmltvChannel,
  XmltvCreditImage,
  XmltvCredits,
  XmltvDOM,
  XmltvDisplayName,
  XmltvEpisodeNumber,
  XmltvIcon,
  XmltvImage,
  XmltvLength,
  XmltvPerson,
  XmltvPreviouslyShown,
  XmltvProgramme,
  XmltvRating,
  XmltvReview,
  XmltvStarRating,
  XmltvSubtitle,
  XmltvUrl,
  XmltvVideo,
} from "./types";

type ParseXmltvOptions = {
  asDom: boolean;
};

type WriteXmltvOptions = {
  fromDom: boolean;
};

/**
 * parseXmltv
 *
 * Parses an xmltv file and returns an `Xmltv` object or a DOM tree
 *
 * @param xmltvString The xmltv file content as a string
 * @param options Options to parse the xmltv file
 * @param options.asDom If true, the xmltv file will be returned as a DOM tree
 */
function parseXmltv(
  xmltvString: string,
  options: ParseXmltvOptions & { asDom: true }
): XmltvDOM;
function parseXmltv(
  xmltvString: string,
  options: ParseXmltvOptions & { asDom: false }
): XmltvDOM;
function parseXmltv(xmltvString: string): Xmltv;

function parseXmltv(
  xmltvString: string,
  options: ParseXmltvOptions = { asDom: false }
): Xmltv | XmltvDOM {
  const parsed = parser(xmltvString);
  if (options.asDom) {
    return parsed;
  }

  return <Xmltv>toObject(parsed);
}

/**
 * writeXmltv
 *
 * Writes an `Xmltv` object or a DOM tree to an xmltv string
 *
 * @param xmltv The `Xmltv` object or a DOM tree
 * @param options Options to write the xmltv file
 * @param options.fromDom If true, the xmltv file will be written from a DOM tree
 * @returns The xmltv file content as a string
 * @throws If `options.fromDom` is true and `xmltv` is an `Xmltv` object
 */
function writeXmltv(
  xmltv: XmltvDOM,
  options: WriteXmltvOptions & { fromDom: true }
): string;
function writeXmltv(
  xmltv: Xmltv,
  options: WriteXmltvOptions & { fromDom: false }
): string;
function writeXmltv(xmltv: Xmltv): string;

function writeXmltv(
  xmltv: Xmltv | XmltvDOM,
  options: WriteXmltvOptions = { fromDom: false }
): string {
  if (options.fromDom) {
    if (typeof xmltv === "object" && !Array.isArray(xmltv)) {
      throw new Error(
        "Cannot write XMLTV from a DOM object that has been converted to an object"
      );
    }
    return writer(xmltv);
  }
  const dom = objectToDom(xmltv);
  return writer(dom);
}

export { parseXmltv, writeXmltv, writer, objectToDom, parser };

export type {
  Xmltv,
  XmltvChannel,
  XmltvDisplayName,
  XmltvProgramme,
  XmltvAudio,
  XmltvCreditImage,
  XmltvCredits,
  XmltvEpisodeNumber,
  XmltvIcon,
  XmltvImage,
  XmltvLength,
  XmltvPerson,
  XmltvPreviouslyShown,
  XmltvRating,
  XmltvReview,
  XmltvStarRating,
  XmltvSubtitle,
  XmltvUrl,
  XmltvVideo,
};
