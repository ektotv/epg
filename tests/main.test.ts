import { describe, expect, test } from "vitest";
import { objectToDom, parseXmltv, writeXmltv } from "../src/main";
import fs from "node:fs";

const fileAsString = fs.readFileSync(`./tests/fixtures/example.xml`, {
  encoding: "utf-8",
});

const fileWithOnlyChannelData = fs.readFileSync(
  `./tests/fixtures/channel-only.xml`,
  {
    encoding: "utf-8",
  }
);

describe("Parses a XMLTV file", () => {
  test("Fully parses to a JS object", async () => {
    const parsed = parseXmltv(fileAsString);
    expect(parsed).toMatchSnapshot();
  });

  test("it can convert back to xml", () => {
    const parsed = parseXmltv(fileAsString);
    const xml = writeXmltv(parsed);
    expect(xml).toMatchSnapshot();
  });

  test("it can convert an object into the dom structure", () => {
    const obj = parseXmltv(fileAsString);
    const dom = objectToDom(obj);

    expect(dom).toMatchSnapshot();
  });

  test("convertToDom produces the same DOM as an xml file parse", () => {
    const [, , exampleOutput] = parseXmltv(fileAsString, { asDom: true });
    const xml = parseXmltv(fileAsString);
    const [dom] = objectToDom(xml);

    expect(exampleOutput).toEqual(dom);
  });

  test("Handles empty input string gracefully", () => {
    const parsed = parseXmltv("");
    expect(parsed).toEqual({});
  });

  test("Throws error on malformed XML input string", () => {
    const malformedXml = "<tv><channel></tv>";
    expect(() => parseXmltv(malformedXml)).toThrowError(`Unexpected close tag
Line: 0
Column: 18
Char: >`);
  });

  test("Parses a XMLTV file with only channel data", () => {
    const parsed = parseXmltv(fileWithOnlyChannelData);

    expect(parsed).toEqual({
      channels: [{ displayName: [{ _value: "Channel 1" }], id: "1" }],
    });
  });
});
