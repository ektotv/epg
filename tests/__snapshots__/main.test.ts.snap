// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Parses a XMLTV file > Fully parses to a JS object 1`] = `
{
  "channels": [
    {
      "displayName": [
        {
          "_value": "Channel One",
          "lang": "en",
        },
      ],
      "icon": [
        {
          "height": 100,
          "src": "https://example.com/channel_one_icon.jpg",
          "width": 100,
        },
      ],
      "id": "channel_one",
      "url": [
        {
          "_value": "https://example.com/channel_one",
          "system": "example",
        },
        {
          "_value": "https://example.com/channel_one_alternate",
          "system": "other_system",
        },
      ],
    },
    {
      "displayName": [
        {
          "_value": "Channel Two: Minimum valid channel",
        },
      ],
      "id": "channel_two",
    },
  ],
  "date": 2022-04-01T00:00:00.000Z,
  "generatorInfoName": "Example Generator",
  "generatorInfoUrl": "example generator url",
  "programmes": [
    {
      "audio": {
        "present": true,
        "stereo": "Dolby Digital",
      },
      "category": [
        {
          "_value": "Crime",
          "lang": "en",
        },
        {
          "_value": "Drama",
          "lang": "en",
        },
      ],
      "channel": "channel_one",
      "clumpidx": "0/1",
      "country": {
        "_value": "US",
      },
      "credits": {
        "actor": [
          {
            "_value": "David Thompson",
            "role": "Walter Johnson",
          },
          {
            "_value": "Ryan Lee",
            "guest": true,
            "image": [
              {
                "_value": "https://www.example.com/xxx.jpg",
                "type": "person",
              },
            ],
            "role": "Karl James",
            "url": [
              {
                "_value": "https://www.example.com/person/204",
                "system": "moviedb",
              },
            ],
          },
        ],
        "adapter": [
          {
            "_value": "William Brown",
          },
        ],
        "commentator": [
          {
            "_value": "James Wilson",
          },
        ],
        "composer": [
          {
            "_value": "Max Wright",
          },
        ],
        "director": [
          {
            "_value": "Samuel Jones",
          },
        ],
        "editor": [
          {
            "_value": "Nora Peterson",
          },
        ],
        "guest": [
          {
            "_value": "Lucas Martin",
          },
          {
            "_value": "Emily Parker",
          },
          {
            "_value": "Oliver Nelson",
          },
        ],
        "presenter": [
          {
            "_value": "Amanda Johnson",
          },
        ],
        "producer": [
          {
            "_value": "Emily Davis",
          },
        ],
        "writer": [
          {
            "_value": "Samuel Jones",
          },
        ],
      },
      "date": 2022-04-01T00:00:00.000Z,
      "desc": [
        {
          "_value": "This programme entry showcases all possible features of the DTD",
          "lang": "en",
        },
      ],
      "episodeNum": [
        {
          "_value": "S01E01",
          "system": "onscreen",
        },
        {
          "_value": "1.1.",
          "system": "xmltv_ns",
        },
      ],
      "icon": [
        {
          "height": 100,
          "src": "https://example.com/programme_one_icon.jpg",
          "width": 100,
        },
      ],
      "image": [
        {
          "_value": "https://tvdb.com/programme_one_poster_1.jpg",
          "orient": "P",
          "size": 1,
          "system": "tvdb",
          "type": "poster",
        },
        {
          "_value": "https://tvdb.com/programme_one_backdrop_3.jpg",
          "orient": "L",
          "size": 3,
          "system": "tvdb",
          "type": "backdrop",
        },
      ],
      "keyword": [
        {
          "_value": "methamphetamine",
          "lang": "en",
        },
        {
          "_value": "cancer",
          "lang": "en",
        },
      ],
      "language": {
        "_value": "English",
      },
      "lastChance": {
        "_value": "Last time on this channel",
        "lang": "en",
      },
      "length": {
        "_value": 60,
        "units": "minutes",
      },
      "new": true,
      "origLanguage": {
        "_value": "French",
        "lang": "en",
      },
      "pdcStart": 2022-03-31T18:00:00.000Z,
      "premiere": {
        "_value": "First time on British TV",
      },
      "previouslyShown": {
        "channel": "channel_two",
        "start": 2022-03-31T18:00:00.000Z,
      },
      "rating": [
        {
          "icon": [
            {
              "src": "15_symbol.png",
            },
          ],
          "system": "BBFC",
          "value": "15",
        },
      ],
      "review": [
        {
          "_value": "This is a fantastic show!",
          "lang": "en",
          "reviewer": "Joe Bloggs",
          "source": "Rotten Tomatoes",
          "type": "text",
        },
        {
          "_value": "https://example.com/programme_one_review",
          "lang": "en",
          "reviewer": "Joe Bloggs",
          "source": "Rotten Tomatoes",
          "type": "url",
        },
      ],
      "showview": "12345",
      "starRating": [
        {
          "icon": [
            {
              "src": "stars.png",
            },
          ],
          "system": "TV Guide",
          "value": "4/5",
        },
      ],
      "start": 2022-03-31T18:00:00.000Z,
      "stop": 2022-03-31T19:00:00.000Z,
      "subTitle": [
        {
          "_value": "Pilot",
          "lang": "en",
        },
      ],
      "subtitles": [
        {
          "language": {
            "_value": "English",
          },
          "type": "teletext",
        },
        {
          "language": {
            "_value": "Spanish",
            "lang": "en",
          },
          "type": "onscreen",
        },
      ],
      "title": [
        {
          "_value": "Programme One",
          "lang": "en",
        },
      ],
      "url": [
        {
          "_value": "https://example.com/programme_one",
          "system": "tvdb",
        },
        {
          "_value": "https://example.com/programme_one_2",
        },
      ],
      "video": {
        "aspect": "16:9",
        "colour": false,
        "present": true,
        "quality": "HDTV",
      },
      "videoplus": "67890",
      "vpsStart": 2022-03-31T18:00:00.000Z,
    },
    {
      "channel": "channel_one",
      "start": 2022-03-31T18:00:00.000Z,
      "title": [
        {
          "_value": "Programme Two: The minimum valid programme",
        },
      ],
    },
  ],
  "sourceDataUrl": "example.com/a",
  "sourceInfoName": "example",
  "sourceInfoUrl": "example.com",
}
`;

exports[`Parses a XMLTV file > it can convert an object into the dom structure 1`] = `
[
  {
    "attributes": {
      "date": "20220401000000 +0000",
      "generator-info-name": "Example Generator",
      "generator-info-url": "example generator url",
      "source-data-url": "example.com/a",
      "source-info-name": "example",
      "source-info-url": "example.com",
    },
    "children": [
      {
        "attributes": {
          "id": "channel_one",
        },
        "children": [
          {
            "attributes": {
              "lang": "en",
            },
            "children": [
              "Channel One",
            ],
            "tagName": "display-name",
          },
          {
            "attributes": {
              "height": "100",
              "src": "https://example.com/channel_one_icon.jpg",
              "width": "100",
            },
            "children": [],
            "tagName": "icon",
          },
          {
            "attributes": {
              "system": "example",
            },
            "children": [
              "https://example.com/channel_one",
            ],
            "tagName": "url",
          },
          {
            "attributes": {
              "system": "other_system",
            },
            "children": [
              "https://example.com/channel_one_alternate",
            ],
            "tagName": "url",
          },
        ],
        "tagName": "channel",
      },
      {
        "attributes": {
          "id": "channel_two",
        },
        "children": [
          {
            "attributes": {},
            "children": [
              "Channel Two: Minimum valid channel",
            ],
            "tagName": "display-name",
          },
        ],
        "tagName": "channel",
      },
      {
        "attributes": {
          "channel": "channel_one",
          "clumpidx": "0/1",
          "pdc-start": "20220331180000 +0000",
          "showview": "12345",
          "start": "20220331180000 +0000",
          "stop": "20220331190000 +0000",
          "videoplus": "67890",
          "vps-start": "20220331180000 +0000",
        },
        "children": [
          {
            "attributes": {
              "lang": "en",
            },
            "children": [
              "Programme One",
            ],
            "tagName": "title",
          },
          {
            "attributes": {
              "lang": "en",
            },
            "children": [
              "Pilot",
            ],
            "tagName": "sub-title",
          },
          {
            "attributes": {
              "lang": "en",
            },
            "children": [
              "This programme entry showcases all possible features of the DTD",
            ],
            "tagName": "desc",
          },
          {
            "attributes": {},
            "children": [
              {
                "attributes": {},
                "children": [
                  "Samuel Jones",
                ],
                "tagName": "director",
              },
              {
                "attributes": {
                  "role": "Walter Johnson",
                },
                "children": [
                  "David Thompson",
                ],
                "tagName": "actor",
              },
              {
                "attributes": {
                  "guest": "yes",
                  "role": "Karl James",
                },
                "children": [
                  "Ryan Lee",
                  {
                    "attributes": {
                      "type": "person",
                    },
                    "children": [
                      "https://www.example.com/xxx.jpg",
                    ],
                    "tagName": "image",
                  },
                  {
                    "attributes": {
                      "system": "moviedb",
                    },
                    "children": [
                      "https://www.example.com/person/204",
                    ],
                    "tagName": "url",
                  },
                ],
                "tagName": "actor",
              },
              {
                "attributes": {},
                "children": [
                  "Samuel Jones",
                ],
                "tagName": "writer",
              },
              {
                "attributes": {},
                "children": [
                  "William Brown",
                ],
                "tagName": "adapter",
              },
              {
                "attributes": {},
                "children": [
                  "Emily Davis",
                ],
                "tagName": "producer",
              },
              {
                "attributes": {},
                "children": [
                  "Max Wright",
                ],
                "tagName": "composer",
              },
              {
                "attributes": {},
                "children": [
                  "Nora Peterson",
                ],
                "tagName": "editor",
              },
              {
                "attributes": {},
                "children": [
                  "Amanda Johnson",
                ],
                "tagName": "presenter",
              },
              {
                "attributes": {},
                "children": [
                  "James Wilson",
                ],
                "tagName": "commentator",
              },
              {
                "attributes": {},
                "children": [
                  "Lucas Martin",
                ],
                "tagName": "guest",
              },
              {
                "attributes": {},
                "children": [
                  "Emily Parker",
                ],
                "tagName": "guest",
              },
              {
                "attributes": {},
                "children": [
                  "Oliver Nelson",
                ],
                "tagName": "guest",
              },
            ],
            "tagName": "credits",
          },
          {
            "attributes": {},
            "children": [
              "20220401000000 +0000",
            ],
            "tagName": "date",
          },
          {
            "attributes": {
              "lang": "en",
            },
            "children": [
              "Crime",
            ],
            "tagName": "category",
          },
          {
            "attributes": {
              "lang": "en",
            },
            "children": [
              "Drama",
            ],
            "tagName": "category",
          },
          {
            "attributes": {
              "lang": "en",
            },
            "children": [
              "methamphetamine",
            ],
            "tagName": "keyword",
          },
          {
            "attributes": {
              "lang": "en",
            },
            "children": [
              "cancer",
            ],
            "tagName": "keyword",
          },
          {
            "attributes": {},
            "children": [
              "English",
            ],
            "tagName": "language",
          },
          {
            "attributes": {
              "lang": "en",
            },
            "children": [
              "French",
            ],
            "tagName": "orig-language",
          },
          {
            "attributes": {
              "units": "minutes",
            },
            "children": [
              "60",
            ],
            "tagName": "length",
          },
          {
            "attributes": {
              "height": "100",
              "src": "https://example.com/programme_one_icon.jpg",
              "width": "100",
            },
            "children": [],
            "tagName": "icon",
          },
          {
            "attributes": {
              "system": "tvdb",
            },
            "children": [
              "https://example.com/programme_one",
            ],
            "tagName": "url",
          },
          {
            "attributes": {},
            "children": [
              "https://example.com/programme_one_2",
            ],
            "tagName": "url",
          },
          {
            "attributes": {},
            "children": [
              "US",
            ],
            "tagName": "country",
          },
          {
            "attributes": {
              "system": "onscreen",
            },
            "children": [
              "S01E01",
            ],
            "tagName": "episode-num",
          },
          {
            "attributes": {
              "system": "xmltv_ns",
            },
            "children": [
              "1.1.",
            ],
            "tagName": "episode-num",
          },
          {
            "attributes": {},
            "children": [
              {
                "attributes": {},
                "children": [
                  "yes",
                ],
                "tagName": "present",
              },
              {
                "attributes": {},
                "children": [
                  "no",
                ],
                "tagName": "colour",
              },
              {
                "attributes": {},
                "children": [
                  "16:9",
                ],
                "tagName": "aspect",
              },
              {
                "attributes": {},
                "children": [
                  "HDTV",
                ],
                "tagName": "quality",
              },
            ],
            "tagName": "video",
          },
          {
            "attributes": {},
            "children": [
              {
                "attributes": {},
                "children": [
                  "yes",
                ],
                "tagName": "present",
              },
              {
                "attributes": {},
                "children": [
                  "Dolby Digital",
                ],
                "tagName": "stereo",
              },
            ],
            "tagName": "audio",
          },
          {
            "attributes": {
              "channel": "channel_two",
              "start": "20220331180000 +0000",
            },
            "children": [],
            "tagName": "previously-shown",
          },
          {
            "attributes": {},
            "children": [
              "First time on British TV",
            ],
            "tagName": "premiere",
          },
          {
            "attributes": {
              "lang": "en",
            },
            "children": [
              "Last time on this channel",
            ],
            "tagName": "last-chance",
          },
          {
            "attributes": {},
            "children": [],
            "tagName": "new",
          },
          {
            "attributes": {
              "type": "teletext",
            },
            "children": [
              {
                "attributes": {},
                "children": [
                  "English",
                ],
                "tagName": "language",
              },
            ],
            "tagName": "subtitles",
          },
          {
            "attributes": {
              "type": "onscreen",
            },
            "children": [
              {
                "attributes": {
                  "lang": "en",
                },
                "children": [
                  "Spanish",
                ],
                "tagName": "language",
              },
            ],
            "tagName": "subtitles",
          },
          {
            "attributes": {
              "system": "BBFC",
            },
            "children": [
              {
                "attributes": {},
                "children": [
                  "15",
                ],
                "tagName": "value",
              },
              {
                "attributes": {
                  "src": "15_symbol.png",
                },
                "children": [],
                "tagName": "icon",
              },
            ],
            "tagName": "rating",
          },
          {
            "attributes": {
              "system": "TV Guide",
            },
            "children": [
              {
                "attributes": {},
                "children": [
                  "4/5",
                ],
                "tagName": "value",
              },
              {
                "attributes": {
                  "src": "stars.png",
                },
                "children": [],
                "tagName": "icon",
              },
            ],
            "tagName": "star-rating",
          },
          {
            "attributes": {
              "lang": "en",
              "reviewer": "Joe Bloggs",
              "source": "Rotten Tomatoes",
              "type": "text",
            },
            "children": [
              "This is a fantastic show!",
            ],
            "tagName": "review",
          },
          {
            "attributes": {
              "lang": "en",
              "reviewer": "Joe Bloggs",
              "source": "Rotten Tomatoes",
              "type": "url",
            },
            "children": [
              "https://example.com/programme_one_review",
            ],
            "tagName": "review",
          },
          {
            "attributes": {
              "orient": "P",
              "size": "1",
              "system": "tvdb",
              "type": "poster",
            },
            "children": [
              "https://tvdb.com/programme_one_poster_1.jpg",
            ],
            "tagName": "image",
          },
          {
            "attributes": {
              "orient": "L",
              "size": "3",
              "system": "tvdb",
              "type": "backdrop",
            },
            "children": [
              "https://tvdb.com/programme_one_backdrop_3.jpg",
            ],
            "tagName": "image",
          },
        ],
        "tagName": "programme",
      },
      {
        "attributes": {
          "channel": "channel_one",
          "start": "20220331180000 +0000",
        },
        "children": [
          {
            "attributes": {},
            "children": [
              "Programme Two: The minimum valid programme",
            ],
            "tagName": "title",
          },
        ],
        "tagName": "programme",
      },
    ],
    "tagName": "tv",
  },
]
`;

exports[`Parses a XMLTV file > it can convert back to xml 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><!DOCTYPE tv SYSTEM \\"xmltv.dtd\\"><tv date=\\"20220401000000 +0000\\" source-info-name=\\"example\\" source-info-url=\\"example.com\\" source-data-url=\\"example.com/a\\" generator-info-name=\\"Example Generator\\" generator-info-url=\\"example generator url\\"><channel id=\\"channel_one\\"><display-name lang=\\"en\\">Channel One</display-name><icon src=\\"https://example.com/channel_one_icon.jpg\\" width=\\"100\\" height=\\"100\\"/><url system=\\"example\\">https://example.com/channel_one</url><url system=\\"other_system\\">https://example.com/channel_one_alternate</url></channel><channel id=\\"channel_two\\"><display-name>Channel Two: Minimum valid channel</display-name></channel><programme start=\\"20220331180000 +0000\\" stop=\\"20220331190000 +0000\\" channel=\\"channel_one\\" pdc-start=\\"20220331180000 +0000\\" vps-start=\\"20220331180000 +0000\\" showview=\\"12345\\" videoplus=\\"67890\\" clumpidx=\\"0/1\\"><title lang=\\"en\\">Programme One</title><sub-title lang=\\"en\\">Pilot</sub-title><desc lang=\\"en\\">This programme entry showcases all possible features of the DTD</desc><credits><director>Samuel Jones</director><actor role=\\"Walter Johnson\\">David Thompson</actor><actor role=\\"Karl James\\" guest=\\"yes\\">Ryan Lee<image type=\\"person\\">https://www.example.com/xxx.jpg</image><url system=\\"moviedb\\">https://www.example.com/person/204</url></actor><writer>Samuel Jones</writer><adapter>William Brown</adapter><producer>Emily Davis</producer><composer>Max Wright</composer><editor>Nora Peterson</editor><presenter>Amanda Johnson</presenter><commentator>James Wilson</commentator><guest>Lucas Martin</guest><guest>Emily Parker</guest><guest>Oliver Nelson</guest></credits><date>20220401000000 +0000</date><category lang=\\"en\\">Crime</category><category lang=\\"en\\">Drama</category><keyword lang=\\"en\\">methamphetamine</keyword><keyword lang=\\"en\\">cancer</keyword><language>English</language><orig-language lang=\\"en\\">French</orig-language><length units=\\"minutes\\">60</length><icon src=\\"https://example.com/programme_one_icon.jpg\\" width=\\"100\\" height=\\"100\\"/><url system=\\"tvdb\\">https://example.com/programme_one</url><url>https://example.com/programme_one_2</url><country>US</country><episode-num system=\\"onscreen\\">S01E01</episode-num><episode-num system=\\"xmltv_ns\\">1.1.</episode-num><video><present>yes</present><colour>no</colour><aspect>16:9</aspect><quality>HDTV</quality></video><audio><present>yes</present><stereo>Dolby Digital</stereo></audio><previously-shown start=\\"20220331180000 +0000\\" channel=\\"channel_two\\"/><premiere>First time on British TV</premiere><last-chance lang=\\"en\\">Last time on this channel</last-chance><new/><subtitles type=\\"teletext\\"><language>English</language></subtitles><subtitles type=\\"onscreen\\"><language lang=\\"en\\">Spanish</language></subtitles><rating system=\\"BBFC\\"><value>15</value><icon src=\\"15_symbol.png\\"/></rating><star-rating system=\\"TV Guide\\"><value>4/5</value><icon src=\\"stars.png\\"/></star-rating><review type=\\"text\\" source=\\"Rotten Tomatoes\\" reviewer=\\"Joe Bloggs\\" lang=\\"en\\">This is a fantastic show!</review><review type=\\"url\\" source=\\"Rotten Tomatoes\\" reviewer=\\"Joe Bloggs\\" lang=\\"en\\">https://example.com/programme_one_review</review><image type=\\"poster\\" size=\\"1\\" orient=\\"P\\" system=\\"tvdb\\">https://tvdb.com/programme_one_poster_1.jpg</image><image type=\\"backdrop\\" size=\\"3\\" orient=\\"L\\" system=\\"tvdb\\">https://tvdb.com/programme_one_backdrop_3.jpg</image></programme><programme start=\\"20220331180000 +0000\\" channel=\\"channel_one\\"><title>Programme Two: The minimum valid programme</title></programme></tv>"`;
